### ISSUE TITLE
Adding Compass to a directory throws an error

### ISSUE THREAD

**Issue by mgrn0** [bdkjones/CodeKit2#11]
Happens when creating a new compass project as well as opening an old (Codekit v1) compass based project.
 Log outputs the following error:
![2013-12-18_12-22-35](https://f.cloud.github.com/assets/3019226/1772718/edc39928-67d6-11e3-84eb-f3462d44e351.jpg)

OSX is 10.8.5, tried locally (not on remote disc or anything).
Not with libsass compiler (option disabled).

When I try to add/install compass to the empty directory, the error is thrown and the folder stays empty, besides the codekit config file.

**Comment by mgrn0** [comment]
addition: tried to create the new project on the Desktop, no weird path or pathname including special characters.

**Comment by mgrn0** [comment]
Ruby is 1.8.7, default installed with OSX 10.8.. can this be the issue / culprit?

ruby 1.8.7 (2012-02-08 patchlevel 358) [universal-darwin12.0]

**Comment by tombonez** [comment]
Having the same issue here, running Mavericks:

```
Compass was unable to compile one or more files in the project: 

/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/rubygems/core_ext/kernel_require.rb:45:in `require': cannot load such file -- listen (LoadError)
    from /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/rubygems/core_ext/kernel_require.rb:45:in `require'
    from /Applications/CodeKit.app/Contents/Resources/engines/compass/lib/compass/watcher/project_watcher.rb:1:in `<top (required)>'
    from /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/rubygems/core_ext/kernel_require.rb:45:in `require'
    from /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/rubygems/core_ext/kernel_require.rb:45:in `require'
    from /Applications/CodeKit.app/Contents/Resources/engines/compass/lib/compass/watcher.rb:3:in `<top (required)>'
    from /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/rubygems/core_ext/kernel_require.rb:45:in `require'
    from /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/rubygems/core_ext/kernel_require.rb:45:in `require'
    from /Applications/CodeKit.app/Contents/Resources/engines/compass/lib/compass.rb:33:in `block in <top (required)>'
    from /Applications/CodeKit.app/Contents/Resources/engines/compass/lib/compass.rb:32:in `each'
    from /Applications/CodeKit.app/Contents/Resources/engines/compass/lib/compass.rb:32:in `<top (required)>'
    from /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/rubygems/core_ext/kernel_require.rb:45:in `require'
    from /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/rubygems/core_ext/kernel_require.rb:45:in `require'
    from /Applications/CodeKit.app/Contents/Resources/engines/compass/bin/compass:32:in `block in <main>'
    from /Applications/CodeKit.app/Contents/Resources/engines/compass/bin/compass:20:in `fallback_load_path'
    from /Applications/CodeKit.app/Contents/Resources/engines/compass/bin/compass:31:in `<main>'
```

**Comment by sethta** [comment]
I also have the same issue with Mavericks (10.9.1)

```
Compass was unable to compile one or more files in the project: 

/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/rubygems/core_ext/kernel_require.rb:45:in `require': cannot load such file -- listen (LoadError)
    from /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/rubygems/core_ext/kernel_require.rb:45:in `require'
    from /Applications/CodeKit copy.app/Contents/Resources/engines/compass/lib/compass/watcher/project_watcher.rb:1:in `<top (required)>'
    from /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/rubygems/core_ext/kernel_require.rb:45:in `require'
    from /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/rubygems/core_ext/kernel_require.rb:45:in `require'
    from /Applications/CodeKit copy.app/Contents/Resources/engines/compass/lib/compass/watcher.rb:3:in `<top (required)>'
    from /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/rubygems/core_ext/kernel_require.rb:45:in `require'
    from /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/rubygems/core_ext/kernel_require.rb:45:in `require'
    from /Applications/CodeKit copy.app/Contents/Resources/engines/compass/lib/compass.rb:33:in `block in <top (required)>'
    from /Applications/CodeKit copy.app/Contents/Resources/engines/compass/lib/compass.rb:32:in `each'
    from /Applications/CodeKit copy.app/Contents/Resources/engines/compass/lib/compass.rb:32:in `<top (required)>'
    from /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/rubygems/core_ext/kernel_require.rb:45:in `require'
    from /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/rubygems/core_ext/kernel_require.rb:45:in `require'
    from /Applications/CodeKit copy.app/Contents/Resources/engines/compass/bin/compass:32:in `block in <main>'
    from /Applications/CodeKit copy.app/Contents/Resources/engines/compass/bin/compass:20:in `fallback_load_path'
    from /Applications/CodeKit copy.app/Contents/Resources/engines/compass/bin/compass:31:in `<main>'
```

**Comment by mdavis1982** [comment]
@tombonez @sethta Can you try installing compass locally with the `--pre` option to get the latest beta, and see if that also throws the same error on the command line please? If so, it's an issue with that as CodeKit has the latest beta included by default.

**Comment by sethta** [comment]
Also, my error happened when I tried to save a scss file, not when I tried to create a new project. I can also confirm that adding a Compass project throws the same error. Just closed CK2 and opened CK1. No issues with Compass.

**Comment by sethta** [comment]
@mdavis1982 I'll give that a try and let you know.

**Comment by sethta** [comment]
I can't even get the `--pre` option installed. It keeps throwing an error, so I can unfortunately test to latest beta. There's no problem installing it locally without the `--pre` option.

```
ERROR:  Error installing compass:
    ERROR: Failed to build gem native extension.

    /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/bin/ruby extconf.rb
mkmf.rb can't find header files for ruby at /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/include/ruby.h


Gem files will remain installed in /Library/Ruby/Gems/2.0.0/gems/ffi-1.9.3 for inspection.
Results logged to /Library/Ruby/Gems/2.0.0/gems/ffi-1.9.3/ext/ffi_c/gem_make.out
```

**Comment by mdavis1982** [comment]
I'd suggest that this is an error with Ruby or the pre release version of SASS then, rather than a CodeKit issue :smiley_cat:

**Comment by mdavis1982** [comment]
@sethta What do you get if you do `ruby -v` on the command line?

**Comment by sethta** [comment]
`ruby 2.0.0p247 (2013-06-27 revision 41674) [universal.x86_64-darwin13]`

**Comment by mdavis1982** [comment]
@sethta Have you install that version of Ruby yourself or is it the one that came with Mavericks?

**Comment by sethta** [comment]
It's the one that came with Mavericks. With the exception of Compass, I don't do anything with Ruby so I have never had a need to try to install it myself.

**Comment by frankyonnetti** [comment]
Same exact issue for me too.

**Comment by maxlapides** [comment]
I'll jump in, too. I'm having the same issue as @tombonez and @sethta reported above.

**Comment by mdavis1982** [comment]
This is definitely related to the pre release version of SASS / Compass - if it won't install on the command line, it probably isn't going to work properly in CodeKit either.

**Comment by mgrn0** [comment]
what about us guys in 10.8 ? will codekit 2 work properly with compass? as stated very above (I opened this issue) I am no on Mavericks yet, my Ruby is the system default (1.8.7 (2012-02-08 patchlevel 358) [universal-darwin12.0]). Can I help somehow to test to get it running on 10.8.5  (latest updates applied).

**Comment by tombonez** [comment]
Quick update, not sure if updating RubyGems (I'd guess so) or installing Compass pre fixed the issue but I'm now up and running (using the internal compiler).

```
gem update --system
gem install compass --pre
```

**Comment by mgrn0** [comment]
Tried the update at first, did not help (on 10.8.5), error stays. will try your suggestion with pre fixed compass.

**Comment by mgrn0** [comment]
no success, codekit v1 still happily compiling, though..

**Comment by mdavis1982** [comment]
@mgrn0 Can you try doing:

``` shell
sudo gem update --system
sudo gem install compass --pre
```

And then seeing if it works? Thanks

**Comment by mgrn0** [comment]
first command went through without a hitch ("Latest version currently installed. Aborting.")
Second "delivered":

```
Building native extensions.  This could take a while...
ERROR:  Error installing compass:
    ERROR: Failed to build gem native extension.

    /System/Library/Frameworks/Ruby.framework/Versions/1.8/usr/bin/ruby extconf.rb
mkmf.rb can't find header files for ruby at /System/Library/Frameworks/Ruby.framework/Versions/1.8/usr/lib/ruby/ruby.h

Gem files will remain installed in /Library/Ruby/Gems/1.8/gems/ffi-1.9.3 for inspection.
Results logged to /Library/Ruby/Gems/1.8/gems/ffi-1.9.3/ext/ffi_c/gem_make.out
```

any suggestions from here on?

**Comment by Chrisedmo** [comment]
I'm also getting this. I am using an old Codekit 1 project which was working fine. And now when I save my SCSS file i get the same error as above.

Also running Mavericks & Sublime Text 2, with pages being served on an external port via Jekyll.

I have tried as per @mdavis1982 comment: https://github.com/bdkjones/codekit2beta/issues/11#issuecomment-30917296 But the first command worked fine & updated. The second timed out for me? (error below)

```
ERROR:  While executing gem ... (Gem::RemoteFetcher::FetchError)
    Errno::ETIMEDOUT: Operation timed out - connect(2) (https://rubygems.org/quick/Marshal.4.8/compass-1.0.0.alpha.17.gemspec.rz)
```

**Comment by rpkoller** [comment]
@mgrn0  hmmm i am not entirely sure. searched the requirements for the latest alpha and rc versions of sass and compass but no luck. my guess would be that either sass, compass or even both  of the prerelease versions need a  version > 1.8.7 (which is the shipped version in 10.8) - i can vaguely remember that for some gems 1.9 was necessary. 

the other option would be that the sass and compass version installed afterwards aren't really compatible. in the case you only tried to install compass alongside a present sass install that might be also the issue. maybe uninstall sass and compass both  and then reinstall just compass directly via sudo gem install compass --pre . compass should initialize the install of the sass version it depends on too. 

in the case the issue is due to the first mentioned ruby version you would have to install rvm and then install version 1.9.x alongside the default 1.8.x

hope that helps. best regards Ralf

**Comment by Chrisedmo** [comment]
Think i've cracked it… I _upgraded_ to Mavericks and I had rvm installed, and my mac was using ruby 1.9.3 and _not_ the 2.0.0 that mavericks ships with, so i ran 

```
    rvm system
```

and this makes the computer use ruby 2.0.0, i could then run the compass --pre command as per https://github.com/bdkjones/codekit2beta/issues/11#issuecomment-30917296 , then I restarted codekit - saved my files and boom. Back in business, it compiled. If @mgrn0 & @sethta can try this and feedback?

**Comment by mgrn0** [comment]
ok, I succeeded now ith installing the --pre version. I had to install XCode and their Command line tools (Preferences -> Downloads -> Components and install the Command Line tools).

But Codekit2 still fails to add Compass to a project. damn..
will try to update ruby now.

**Comment by mgrn0** [comment]
I upgraded my machine to Mavericks, did

```
sudo gem update --system
sudo gem install compass --pre
```

but still got the same error.. Ruby is 2.0.0p247 (2013-06-27 rev 41674)

```
ERROR:  Error installing compass:
    ERROR: Failed to build gem native extension.
```

Then re-installed XCode (now Mavericks version), re-tried to install compass --pre (really took a while ;) and that when through without an error.

Codekit 2 now works for me, installing compass in a project works too.

**Comment by sethta** [comment]
@Chrisedmo I managed to get it working now for me. I already had XCode installed, but hadn't opened it since installing it on my new computer, and it hadn't installed any of the command line tools as @mgrn0 mentioned.

All is working well for me now. So I guess the bug is that there is currently some discrepancy with a library somewhere that requires the latest version of XCode and its command line tools to be installed if working with Compass.

**Comment by mgrn0** [comment]
On my second machine I found a _very_ short way of getting compass to run with codekit v2:
- upgraded to Mavericks
- installed _only_ the command line developer tools, NO Xcode required. On terminal: 
  `
  xcode-select --install
  `
-  `sudo gem update --system`
  and
  `sudo gem install compass --pre`

after that, codekit 2 runs smoothly with compass, and no XCode installation required. Took less ten 10 mins (Mavericks update not counted). wohoow!! :)

**Comment by ChristiaanKZ** [comment]
@mgrn0 Thx this does the trick!

**Comment by frankyonnetti** [comment]
This worked for me too, thanks @mgrn0 and @mdavis1982 

`sudo gem update --system` and `sudo gem install compass --pre`

**Comment by mgrn0** [comment]
I got Codekit 2 compiling all my sass and compass projects without installing it with the --pre option, I pointed both sass and compass compilers directly to the bin files, like
![2013-12-30_10-40-02](https://f.cloud.github.com/assets/3019226/1821549/055319ba-7137-11e3-9808-24a8d405e7e4.png)
![2013-12-30_10-39-48](https://f.cloud.github.com/assets/3019226/1821551/0b0336c4-7137-11e3-9b9f-4179287dab2e.png)
(I uninstalled sass, compass and all modules throug  the terminal before).
Has somebody suceeded with this, too? I had to do this because the --pre compass version did not like some modules, like modular-scale, that I require in my sass stylesheets..

**Comment by frankyonnetti** [comment]
Yes @mgrn0, I had to do the same thing, which works. The latest preview of Compass drop some legacy support which I'm using in a few Drupal projects.

@mdavis1982, I'm confused as to why CK2 need the latest Compass --pre to work? When using CK2's internal version of SASS and Compass, why does it care what version of Compass I have installed on my machine. Is it maybe related to the version of Ruby I have locally?

`ruby 2.0.0p247 (2013-06-27 revision 41674) [universal.x86_64-darwin13]`

**Comment by bdkjones** [comment]
CodeKit requires the latest version of Compass because only the --pre release has support for the next version of Sass.

**Comment by chriscoyier** [comment]
Just to throw in my 2 cents, this was difficult for me too. I had Ruby 1.9.3 installed for whatever reason, even after Mavericks. I don't love screwing around with Ruby because I have other projects that use it that are actually Ruby projects, not just things that use Sass. They use `rbenv` though, so I was able to kinda stop using that for a minute, upgrade the system level Ruby stuff, which got it working. Now in my .bash_profile I have this line I need to alternate between commenting out and un-commenting out to either get by Ruby projects working or my CodeKit stuff working, which isn't that huge of a deal especially because once they are up and running they can both work simultaneously.

It's just an interesting situation because I think that' WHY CodeKit is so popular, is this command line trickery crap sucks and people don't like dicking with it so they use a nice app instead.

**Comment by bdkjones** [comment]
Yea. 

The issue is UTF-8 text encoding. Older versions of Ruby established that text encoding as the default, automatically. For whatever completely retarded reason, in Ruby 2.0 they decided to NOT set a default text encoding anymore. This causes every Ruby-based compiler (Sass, Compass, Haml, Slim) to completely fail if a non-ASCII character is used.

So, CodeKit attempts to set the Ruby ENV variable that establishes UTF-8 text encoding. But Ruby being Ruby, the command to do this is only valid in Ruby 2.0. In earlier versions of Ruby, it causes an exception. So CodeKit has to determine which version of Ruby is installed and attempt to react accordingly. This works unless you've customized Ruby in some way. 

 **_TL;DR: Ruby is an evil, shitty language and the world would have been much better off if it had never been written.**_

The other option I have is to not attempt to set the Ruby env variable and, instead, make the user add the appropriate tag to their file. For Sass, it's the usual `@charset "utf-8";` but some languages like HAML and SLIM don't have an equivalent.

**Comment by bdkjones** [comment]
Fixed in Beta 9. Will post shortly.
